mhs/inbound/inbound/request/handler.py

```diff
@@ -2,97 +2,155 @@
+            log.interaction_id.set(interaction_id)
+
+            ref_to_message_id = self._extract_ref_message(request_message)
+            correlation_id = self._extract_correlation_id(request_message)
+            self._extract_message_id(request_message)
+
+            received_message = request_message.message_dictionary[MESSAGE]
+
+            try:
+                work_description = await wd.get_work_description_from_store(self.work_description_store, ref_to_message_id)
+            except wd.EmptyWorkDescriptionError as e:
+                await self._handle_no_work_description_found_for_request(e, ref_to_message_id, correlation_id,
+                                                                         request_message, received_message)
+                return
+
+            message_workflow = self.workflows[work_description.workflow]
+            logger.info('004', 'Retrieved work description from state store, forwarding message to {workflow}',
+                        {'workflow': message_workflow})
+
+            try:
+                await message_workflow.handle_inbound_message(ref_to_message_id, correlation_id, work_description,
+                                                              received_message)
+                self._send_ack(request_message)
+            except Exception as e:
+                logger.error('006', 'Exception in workflow {exception}', {'exception': e})
+                raise tornado.web.HTTPError(500, 'Error occurred during message processing,'
+                                                 ' failed to complete workflow',
+                                            reason=f'Exception in workflow') from e
+
+    async def _handle_no_work_description_found_for_request(self, e: wd.EmptyWorkDescriptionError,
+                                                            ref_to_message_id: str, correlation_id: str,
+                                                            request_message:
+                                                            ebxml_request_envelope.EbxmlRequestEnvelope,
+                                                            received_message: str):
+        # Lookup workflow for request
         interaction_id = request_message.message_dictionary[ebxml_envelope.ACTION]
-        log.interaction_id.set(interaction_id)
+        interaction_details = self._get_interaction_details(interaction_id)
+        message_workflow = self._extract_default_workflow(interaction_details, interaction_id)
 
-        ref_to_message_id = self._extract_ref_message(request_message)
-        correlation_id = self._extract_correlation_id(request_message)
-        self._extract_message_id(request_message)
+        # If it matches forward reliable workflow, then this will be an unsolicited request from another GP system.
+        # So let the workflow handle this.
+        if isinstance(message_workflow, forward_reliable.AsynchronousForwardReliableWorkflow):
```

Comment:
TODO-create techdebt for changing this so that we don't do the isinstance check here and instead let the workflows decide how to handle the unsolicited inbound request

Reply:
Created [RT-298](https://gpitbjss.atlassian.net/browse/RT-298)